### PR TITLE
Add child types to opcode property table

### PR DESCRIPTION
**Pull request by Leonardo2718**
As a first step to resolve issue #1298, this change adds a field to the ILOpCodeProperties table that specifies the expected number and types of child nodes for each opcode. This information should
eventually be used by the tree verifier to ensure that IL is well-formed.

Although I did my best to fill the table with appropriate values, there are many opcodes I am unfamiliar with and therefore had to make educated guesses or leave the `childProperties` field as unspecified.

I will wait a little before removing the WIP to allow time for downstream projects to update their tables.

In the mean time, I would appreciate reviews from @0xdaryl, @vijaysun-omr, @mstoodle, @andrewcraik, @fjeremic, and @ymanton as well as whoever else people think should be involved.

### CODE REVIEW

**Comment by mgaudet** [comment]
Quick comment: Shouldn't we have a `static_assert` somewhere, to ensure a `uint32_t` is sufficiently wide to hold the four pieces of info?

**Comment by mgaudet** [comment]
Oh, and one more general comment: You provide macros as "setters" for the table; Perhaps the PR should also include a "getter" interface?

**Comment by Leonardo2718** [comment]
@mgaudet 

There should most certainly be a `static_assert` somewhere! I'm not convinced however that ensuring a `uint32_t` is sufficiently wide is the correct thing to check.

As it stands, the 32-bit integer is divided into 4, 8-bit segments. One segment is used to store the expected number of child nodes, the other three store the expected type of these nodes. This is already a limitation as it's impossible to represent the expected child types of opcodes with 5 children. However, one thing that can be statically checked is that all possible types can be represented using only 8-bits (remembering that 0xff is a special value).

"Getters" for these values are also coming.

**Comment by Leonardo2718** [comment]
I should make a note that I'm quite wiling to use a `uint64_t` instead, if the added memory cost is considered acceptable.

**Comment by andrewcraik** [comment]
What memory usage evaluation has been done on this change? I think knowing what the usage change is will help guide how aggressive we need to be in factoring this change to minimize growth if that makes sense. The fundamental nature of node and opcode means we do need to be careful that we are doing the right thing. Making the child types more obvious and codified is a good thing.  How are we proposing to handle the opcodes with variable shapes like arraycopy for the time being since factoring the opcode apart is a much bigger change which we may want to do eventually but aren't going to do so as part of this and may not get to for a bit due to other pressing issues needing dev attention?

**Comment by Leonardo2718** [comment]
@andrewcraik 

I haven't had a chance to evaluate memory usage yet. My suspicion is that this change will only impact the final size of the binary because, as far as I know, the properties table is allocated statically and no copies of it are ever made (also, nothing is using the new field yet). That said, I'm open to putting this hypothesis to the test if someone can propose a good evaluation method.

For opcodes like `arraycopy`, I'm leaving them as `ILChildProp::Unspecified` for now. A special case can be added to the tree verifier that will handle these opcodes specifically. Although this is not ideal, I think it's an acceptable solution until the factoring occurs.

**Comment by andrewcraik** [comment]
Ok - given that you expect the growth to be in binary size can you report the growth you have seen in your builds? That at least gives us a starting point for the footprint conversation @mpirvu FYI

**Comment by mpirvu** [comment]
I did a rough calculation for the extra footprint: 731 entries of 4 bytes each ==> about 2.9KB of extra footprint. This is tolerable given that currently the size of the JIT dll is measured in MBs (8MB+)

**Comment by fjeremic** [comment]
Do we know which IL opcode has the largest number of children. Do we exceed 7? I would be in favor of using a 64-bit integer as there are plenty of IL opcodes with > 3 children and the types of such IL opcodes is not captured. This change also sets the basis for starting to Doxygen comment all of our IL.

**Comment by Leonardo2718** [comment]
@fjeremic 

I don't know of any opcode with a fixed number of children that exceeds 7. Maybe someone else can confirm? ( @andrewcraik, @vijaysun-omr, @0xdaryl, @mstoodle ?). I am not counting opcodes that have a variable number of children (e.g. calls).

Following @mpirvu's calculation, using a 64-bit integer would add about 5.8KB of extra footprint (I'm still working on collecting data from local builds). If others agree that this is acceptable (+1), I will change the type of the field to `uint64_t`.

**Comment by vijaysun-omr** [comment]
To my knowledge no IL opcode has a fixed number of children that exceeds 7. I would agree with Marius that the small number of KBs increase you are talking about should be acceptable.

**Comment by Leonardo2718** [comment]
So, oddly enough, adding this field to the opcode table has _decreased_ the final size of the 
JitBuilder static archive. Builds on my local machine are:

* unchanged: 17.31MB
* with the new field: 17.28MB

Here is the output from bloaty with `uint64_t` (note the `FILE SIZE` on the right hand side):

```
./bloaty -d symbols libjitbuilder_treeVerifier.a -- libjitbuilder_master.a
     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
   +10% +7.46Ki _GLOBAL__sub_I_OMRILOps.cpp                                                      +7.46Ki   +10%
   +14% +5.62Ki OMR::ILOpCode::_opCodeProperties                                                       0  [ = ]
  +0.0%    +151 [Unmapped]                                                                          +744  +0.0%
  +2.6%    +713 TR_RegisterCandidates::assign                                                       +713  +2.6%
  +2.6%    +152 TR_Debug::print                                                                     +152  +2.6%
  +6.0%     +97 TR_LoopUnroller::addEdgeAndFixEverything                                             +97  +6.0%
  +1.6%     +96 TR_Debug::printWithFixedPrefix                                                       +96  +1.6%
  +1.2%     +80 TR_LoopReplicator::addBlocksAndFixEdges                                              +80  +1.2%
  +0.7%     +73 [Other]                                                                              +73  +0.7%
  +1.2%     +41 TR_GlobalRegisterAllocator::prepareForBlockExit                                      +41  +1.2%
  +8.5%     +30 TR_LocalLiveRangeReduction::isWorthMoving                                            +30  +8.5%
  +1.6%     +28 TR_LoopCanonicalizer::rewritePostToPreIncrementTestInBlock                           +28  +1.6%
  +0.7%     +27 TR_HoistBlocks::process                                                              +27  +0.7%
  +2.2%     +24 searchForStringAppend                                                                +24  +2.2%
  +7.6%     +17 OMR::Block::getNumberOfRealTreeTops                                                  +17  +7.6%
  +2.2%     +16 treetopSimplifier                                                                    +16  +2.2%
  +0.5%     +16 TR_LocalTransparency::adjustInfoForAddressAdd                                        +16  +0.5%
  +1.0%     +16 OMR::LocalCSE::transformBlock                                                        +16  +1.0%
  +2.0%      +9 TR_RegionStructure::computeInvariantExpressions                                       +9  +2.0%
  +3.8%      +8 findTreeTop                                                                           +8  +3.8%
  +0.3%      +8 lmaxminSimplifier                                                                     +8  +0.3%

 -------------- SHRINKING                                                                        --------------
  -1.7% -32.5Ki [Other]                                                                          -32.5Ki  -1.7%
  -5.6%    -353 addSimplifierCommon                                                                 -353  -5.6%
  -3.5%    -342 TR_PartialRedundancy::placeComputationsOptimally                                    -342  -3.5%
  -2.4%    -317 TR_LoopVersioner::buildBoundCheckComparisonsTree                                    -317  -2.4%
  -1.6%    -215 TR::DeadTreesElimination::process                                                   -215  -1.6%
  -0.7%    -211 TR_ShrinkWrap::doPlacement                                                          -211  -0.7%
  -2.8%    -191 TR_SimplifyAnds::process                                                            -191  -2.8%
  -5.8%    -180 TR_PartialRedundancy::eliminateRedundantComputations                                -180  -5.8%
  -5.6%    -170 OMR::CodeGenerator::identifyUnneededByteConvNodes                                   -170  -5.6%
  -3.0%    -168 constrainIfcmplessthan(OMR::ValuePropagation*, TR::Node*, TR::Node*, TR::Node*,     -168  -3.0%
  -2.0%    -164 TR_LocalTransparency::updateUsesAndDefs                                             -164  -2.0%
  -3.2%    -160 switchSimplifier                                                                    -160  -3.2%
  -1.7%    -160 TR_Rematerialization::examineNode                                                   -160  -1.7%
  -3.9%    -160 TR_ExceptionCheckMotion::composeLists                                               -160  -3.9%
  -4.1%    -160 TR_Debug::printNodeInfo                                                             -160  -4.1%
  -7.0%    -151 TR_LoopVersioner::detectInvariantBoundChecks                                        -151  -7.0%
  -2.5%    -146 TR_GlobalRegisterAllocator::markAutosUsedIn                                         -146  -2.5%
  -5.5%    -144 directStoreSimplifier                                                               -144  -5.5%
  -1.0%    -144 TR_CopyPropagation::perform                                                         -144  -1.0%
  -4.5%    -140 TR_LoadExtensions::detectUnneededConversionPattern                                  -140  -4.5%
  -3.0%    -139 TR_ExceptionCheckMotion::includeRelevantNodes                                       -139  -3.0%

  -0.3% -21.6Ki TOTAL                                                                            -26.6Ki  -0.2%
```

And here is the output from bloaty with `uint32_t`:

```
./bloaty -d symbols libjitbuilder_treeVerifier.a -- libjitbuilder_master.a
     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
   +14% +5.62Ki OMR::ILOpCode::_opCodeProperties                                                       0  [ = ]
  +0.0%    +191 [Unmapped]                                                                          +764  +0.0%
  +2.6%    +713 TR_RegisterCandidates::assign                                                       +713  +2.6%
  +0.6%    +432 _GLOBAL__sub_I_OMRILOps.cpp                                                         +432  +0.6%
  +2.6%    +152 TR_Debug::print                                                                     +152  +2.6%
  +6.0%     +97 TR_LoopUnroller::addEdgeAndFixEverything                                             +97  +6.0%
  +1.6%     +96 TR_Debug::printWithFixedPrefix                                                       +96  +1.6%
  +1.2%     +80 TR_LoopReplicator::addBlocksAndFixEdges                                              +80  +1.2%
  +0.7%     +73 [Other]                                                                              +73  +0.7%
  +1.2%     +41 TR_GlobalRegisterAllocator::prepareForBlockExit                                      +41  +1.2%
  +8.5%     +30 TR_LocalLiveRangeReduction::isWorthMoving                                            +30  +8.5%
  +1.6%     +28 TR_LoopCanonicalizer::rewritePostToPreIncrementTestInBlock                           +28  +1.6%
  +0.7%     +27 TR_HoistBlocks::process                                                              +27  +0.7%
  +2.2%     +24 searchForStringAppend                                                                +24  +2.2%
  +7.6%     +17 OMR::Block::getNumberOfRealTreeTops                                                  +17  +7.6%
  +2.2%     +16 treetopSimplifier                                                                    +16  +2.2%
  +0.5%     +16 TR_LocalTransparency::adjustInfoForAddressAdd                                        +16  +0.5%
  +1.0%     +16 OMR::LocalCSE::transformBlock                                                        +16  +1.0%
  +2.0%      +9 TR_RegionStructure::computeInvariantExpressions                                       +9  +2.0%
  +3.8%      +8 findTreeTop                                                                           +8  +3.8%
  +0.3%      +8 lmaxminSimplifier                                                                     +8  +0.3%

 -------------- SHRINKING                                                                        --------------
  -1.7% -32.5Ki [Other]                                                                          -32.5Ki  -1.7%
  -5.6%    -353 addSimplifierCommon                                                                 -353  -5.6%
  -3.5%    -342 TR_PartialRedundancy::placeComputationsOptimally                                    -342  -3.5%
  -2.4%    -317 TR_LoopVersioner::buildBoundCheckComparisonsTree                                    -317  -2.4%
  -1.6%    -215 TR::DeadTreesElimination::process                                                   -215  -1.6%
  -0.7%    -211 TR_ShrinkWrap::doPlacement                                                          -211  -0.7%
  -2.8%    -191 TR_SimplifyAnds::process                                                            -191  -2.8%
  -5.8%    -180 TR_PartialRedundancy::eliminateRedundantComputations                                -180  -5.8%
  -5.6%    -170 OMR::CodeGenerator::identifyUnneededByteConvNodes                                   -170  -5.6%
  -3.0%    -168 constrainIfcmplessthan(OMR::ValuePropagation*, TR::Node*, TR::Node*, TR::Node*,     -168  -3.0%
  -2.0%    -164 TR_LocalTransparency::updateUsesAndDefs                                             -164  -2.0%
  -3.2%    -160 switchSimplifier                                                                    -160  -3.2%
  -1.7%    -160 TR_Rematerialization::examineNode                                                   -160  -1.7%
  -3.9%    -160 TR_ExceptionCheckMotion::composeLists                                               -160  -3.9%
  -4.1%    -160 TR_Debug::printNodeInfo                                                             -160  -4.1%
  -7.0%    -151 TR_LoopVersioner::detectInvariantBoundChecks                                        -151  -7.0%
  -2.5%    -146 TR_GlobalRegisterAllocator::markAutosUsedIn                                         -146  -2.5%
  -5.5%    -144 directStoreSimplifier                                                               -144  -5.5%
  -1.0%    -144 TR_CopyPropagation::perform                                                         -144  -1.0%
  -4.5%    -140 TR_LoadExtensions::detectUnneededConversionPattern                                  -140  -4.5%
  -3.0%    -139 TR_ExceptionCheckMotion::includeRelevantNodes                                       -139  -3.0%

  -0.4% -28.6Ki TOTAL                                                                            -33.6Ki  -0.2%
```

**Comment by fjeremic** [comment]
@Leonardo2718 if these are 64-bit ELF files the compiler likely padded / aligned the structure when it was `uint32_t` so you might not lose any space.

File: compiler/il/ILProps.hpp [hunk 1]
```diff
@@ -47,6 +49,107 @@ namespace ILTypeProp
...
+#define THIRD_CHILD(type)     (type) << 16
+#define FOURTH_CHILD(type)    (type) << 24
+#define FIFTH_CHILD(type)     (type) << 32
+#define SIXTH_CHILD(type)     (type) << 40
+#define SEVENTH_CHILD(type)   (type) << 48
+#define CHILD_COUNT(c)        (c)    << 56
+
+#define ONE_CHILD(type)                                              CHILD_COUNT(1L) | FIRST_CHILD(type)
+#define TWO_CHILD(type1, type2)                                      CHILD_COUNT(2L) | FIRST_CHILD(type1) | SECOND_CHILD(type2)
+#define THREE_CHILD(type1, type2, type3)                             CHILD_COUNT(3L) | FIRST_CHILD(type1) | SECOND_CHILD(type2) | THIRD_CHILD(type3)
```
Comment (0xdaryl) [inline compiler/il/ILProps.hpp]: Why aren't `FOUR_CHILD` and `SIX_CHILD` defined?  Even if there may be no opcodes that need this many children I think its good to provide them for completeness.

File: compiler/il/ILProps.hpp [hunk 2]
```diff
@@ -47,6 +49,107 @@ namespace ILTypeProp
...
    }
 
+/**
+ * @brief A type representing the expected number and types of child nodes
+ *
+ * The the expected number of children and their types is stored a 64-bit integer.
+ * The 64-bit integer is divided into 8 segments, 8 bits each. The high 8 bits
+ * specify the expected number of children. The remaining bits, from high to low,
+ * specify the expected type of each child, from last to first. Types are
+ * represneted by their TR::DataTypes value. A value of 0xff for any field means
```
Comment (0xdaryl) [inline compiler/il/ILProps.hpp]: Typo: represented

File: compiler/il/ILProps.hpp [hunk 3]
```diff
@@ -47,6 +49,107 @@ namespace ILTypeProp
       };
    }
 
+/**
+ * @brief A type representing the expected number and types of child nodes
+ *
+ * The the expected number of children and their types is stored a 64-bit integer.
```
Comment (Leonardo2718) [inline compiler/il/ILProps.hpp]: Fixed

File: compiler/il/ILProps.hpp (hunk 2 above)
Comment (Leonardo2718) [inline compiler/il/ILProps.hpp]: Fixed.

File: compiler/il/ILProps.hpp (hunk 1 above)
Comment (Leonardo2718) [inline compiler/il/ILProps.hpp]: I've added definitions for `FOUR_CHILD` and `SIX_CHILD`.

**Pull request merged**